### PR TITLE
Replication of metrics (gaps filling) during streaming

### PR DESCRIPTION
**Pull request by vkalintiris**
## Replication

This PR adds support for replicating gaps during streaming of metrics.

It works like this:

1. Child sends chart definition to parent, but does not start replicating metrics. A new child->parent command is introduced `CHART_DEFINITION_END first_time_t last_time_t`  to let the parent know the database boundaries of the child.
2. Parent receives the `CHART_DEFINITION_END` command and sends to the child a `REPLAY_CHART chart.id boolean after before` command. `after` and `before` are adjusted according to user settings `seconds to replicate` (default 86400 seconds, a day) and `replication step` (default 600 seconds).  `boolean` is either `true` or `false` to indicate that the parent wants the child start streaming data after receiving this command.
3. Child receives  `REPLAY_CHART` command and adds it to a queue (dictionary).
4. While the send buffer on the child is less than 50% used, the child picks replication commands from the queue and executes them.
5. For each replication response the child sends `RBEGIN`, `RSET`, `REND` commands to the parent, similarly to the normal streaming. These commands have different parameters compared to normal streaming.
6. If the `boolean` flag to enable streaming is set, the child also transmits `RDSTATE` and `RSSTATE` commands (before the `REND`), to let the parent know of the data collection state parameters, required to allow streaming to seamlessly continue on the parent. Immediately after that, it begins streaming data, like it did before this PR.
7. Until replication is finished, the parent receives replication data and requests the next step.

## Design

- [x] Stateless. The state of replication is on the wire. The parent relies on the child to reply for every single replication request it sends and based on the responses it receives progresses the replication until it is finished.
- [x] Proxy support. There are 2 states related to replication `RECEIVER_FINISHED` and `SENDER_FINISHED`. The receiver finished flag is set on data collecting children and on parents that have finished replication from their children. Chart definitions are sent to parents only when the receiving part has finished replication. This means that a proxy will first replicate data to itself before sends any data to its parent.
- [x] Using a similar protocol to the existing streaming, allows replication to replicate charts with any number of dimensions.
- [x] No gaps after replication completes. The parent and child have the same data.

## Limitations

1. This PR does not replicate metrics that are not currently being collected. Charts will be automatically replicated when they start being collected on a child, even if they started being collected after replication is finished for all other charts. Until charts start being collected, the gaps on the parent will not be filled.
2. The only possibility of data loss is on charts that have a variable number of dimensions and some dimensions are added after the other dimensions of the same chart have been replicated. These dimensions will not replicate their data to the parents.

## Testing

- [x] new child -> new parent (restarting the parent gives no gaps, and all data are perfectly aligned in time)
- [x] old child -> new parent
- [x] new child -> new proxy -> new parent (restarting the proxy gives no gaps on both the proxy and the parent)
- [x] new child (dbengine) -> new proxy (ram) -> new parent (dbengine), replication works without gaps 
- [ ] new child -> old parent

### CODE REVIEW

**Comment by vkalintiris** [comment]
Rebased and force updated. New commits after 0bf815452a573b905e6990587a577aced42f1df2.

@stelfrag this is ready for testing. Replication is enabled by default and needs no extra configuration options to be set manually (we can change this prior to merging the PR).

There's a well known issue where a small gap exists in between the end of the replicated data and the beginning of streaming. This will be fixed in a follow up PR.

**Comment by ktsaou** [comment]
@vkalintiris have you also done the increase of dynamic words count for pluginsd? I don't see it. This means the max dimensions replicated now is 9.

**Comment by vkalintiris** [comment]
> @vkalintiris have you also done the increase of dynamic words count for pluginsd? I don't see it. This means the max dimensions replicated now is 9.

https://github.com/netdata/netdata/pull/13873/files#diff-fa045602da0dc7c7505b55cd2491972b3367e7571a6103f8c4d8195d4cb4fdacR43

We can replicate charts with approx. 100 dims.

The changes required to avoid `memset`ing unused words are [ready](https://github.com/vkalintiris/netdata/tree/replay-words), but I think it better to run a quick second check after testing this PR as it is (because we are touching a lot of things and it's going to be difficult to figure out which change might cause an issue, ie. the replication code or the word splitting).

Do you prefer to do the testing in one go?

**Comment by vkalintiris** [comment]
@Ferroin armv7l & linux/i386 fail at link time. Can you (or let me know how can I) run `file` and `nm` on `streaming/replication.o`?

File: collectors/plugins.d/pluginsd_parser.c [hunk 1]
```diff
@@ -641,68 +667,81 @@ PARSER_RC pluginsd_variable(char **words, void *user, PLUGINSD_ACTION  *plugins_
...
     }
 
     if(unlikely(!((PARSER_USER_OBJECT *) user)->new_host_labels))
         ((PARSER_USER_OBJECT *) user)->new_host_labels = rrdlabels_create();
 
-    rrdlabels_add(((PARSER_USER_OBJECT *)user)->new_host_labels, words[1], store, strtol(words[2], NULL, 10));
+    rrdlabels_add(((PARSER_USER_OBJECT *)user)->new_host_labels,
+                  get_word(words, num_words, 1),
+                  store,
+                  strtol(get_word(words, num_words, 2), NULL, 10));
```
Comment (thiagoftsm) [inline collectors/plugins.d/pluginsd_parser.c:734]: A suggestion, please, change `strtol` to `str2l`.

**Comment by thiagoftsm** [comment]
@vkalintiris ,

Please, take a look on these warnings:

```sh
streaming/replication.cc: In member function 'void ChartQuery::fillLine(BUFFER*, time_t, time_t)':
streaming/replication.cc:142:51: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'int' [-Wformat=]
  142 |                         buffer_sprintf(WB, " %lf %u", SP.sum, SP.flags);
      |                                                  ~^           ~~~~~~~~
      |                                                   |              |
      |                                                   unsigned int   int
      |                                                  %u
collectors/tc.plugin/plugin_tc.c:808: warning: "PLUGINSD_MAX_WORDS" redefined
  808 | #define PLUGINSD_MAX_WORDS 20
      | 
In file included from ./collectors/all.h:10,
                 from ./daemon/common.h:63,
                 from collectors/tc.plugin/plugin_tc.c:3:
./collectors/plugins.d/plugins_d.h:47: note: this is the location of the previous definition
   47 | #define PLUGINSD_MAX_WORDS 512
      | 
libnetdata/libnetdata.c:34:2: warning: #warning NETDATA_TRACE_ALLOCATIONS ENABLED [-Wcpp]
   34 | #warning NETDATA_TRACE_ALLOCATIONS ENABLED
      |  ^~~~~~~
 OK  ''

```

they are generated with the following flags:

```sh
CFLAGS="-Og -ggdb -Wall -Wextra -fno-omit-frame-pointer -Wformat-signedness -fstack-protector-all -Wformat-truncation=2 -Wunused-result -DHAVE_BACKTRACE=1 -DNETDATA_TRACE_ALLOCATIONS=1 -DNETDATA_DEV_MODE=1 -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1"
```

**Comment by ktsaou** [comment]
## Infinite replication (to the future)

```
2022-10-28 22:59:47: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1666987126 1666987726
2022-10-28 22:59:47: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1666987731 1666988331
2022-10-28 22:59:47: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1666988336 1666988936
2022-10-28 22:59:47: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1666988941 1666989541
...
2022-10-28 23:06:36: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1709620871 1709621471
2022-10-28 23:06:36: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1709621476 1709622076
2022-10-28 23:06:36: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1709622081 1709622681
2022-10-28 23:06:36: netdata IERR  : STREAM_SENDER[costa-xps9500] : (0853@streaming/sender.c  :execute_command): STREAM costa-xps9500 [send to 192.168.1.201] received command over connection: REPLAY_CHART "apps.process_create" "false" 1709622686 1709623286
```

Now is:

```sh
# date +%s                                                                                                                                                                                                                                                                          1666987724
```

I killed the child. It continued replicating forever...

**Comment by ktsaou** [comment]
Restarting the child gives this assert:

image

**Comment by ktsaou** [comment]
This is ready to merge.
I have not tested a proxy yet. If someone can help, please do.

File: streaming/sender.c [hunk 2]
```diff
@@ -879,9 +909,32 @@ void execute_commands(struct sender_state *s) {
...
+                      rrdhost_hostname(s->host), s->connected_to,
+                      keyword,
+                      chart_id ? chart_id : "(unset)",
+                      start_streaming ? start_streaming : "(unset)",
+                      after ? after : "(unset)",
+                      before ? before : "(unset)");
+            } else {
+                struct replication_request tmp = {
+                        .start_streaming = !strcmp(start_streaming, "true"),
+                        .after = strtoll(after, NULL, 0),
```
Comment (thiagoftsm) [inline streaming/sender.c:930]: Please, use `str2ll` instead `strtoll` for these two variables.

File: collectors/plugins.d/pluginsd_parser.c [hunk 3]
```diff
@@ -855,6 +899,291 @@ PARSER_RC metalog_pluginsd_host(char **words, void *user, PLUGINSD_ACTION  *plug
...
+        }
+
+        ((PARSER_USER_OBJECT *) user)->st = st;
+        ((PARSER_USER_OBJECT *) user)->replay.start_time = 0;
+        ((PARSER_USER_OBJECT *) user)->replay.end_time = 0;
+        ((PARSER_USER_OBJECT *) user)->replay.start_time_ut = 0;
+        ((PARSER_USER_OBJECT *) user)->replay.end_time_ut = 0;
+    }
+
+    if(start_time_str && end_time_str) {
```
Comment (thiagoftsm) [inline collectors/plugins.d/pluginsd_parser.c:931]: Like previous commit, please, let us use `str2l` here.

File: database/rrdset.c [hunk 4]
```diff
@@ -1895,6 +1862,30 @@ void rrdset_timed_done(RRDSET *st, struct timeval now) {
     netdata_thread_enable_cancelability();
 }
 
+time_t rrdset_set_update_every(RRDSET *st, time_t update_every) {
+
+    internal_error(true, "RRDSET '%s' switching update every from %d to %d",
+                   rrdset_id(st), (int)st->update_every, (int)update_every);
+
+    time_t prev_update_every = st->update_every;
```
Comment (thiagoftsm) [inline database/rrdset.c:1870]: Hello guys, 

A small one from Coverity:

```sh
*** CID 361429:  High impact quality  (Y2K38_SAFETY)
/database/rrdset.c: 1871 in rrdset_set_update_every()
1865     time_t rrdset_set_update_every(RRDSET *st, time_t update_every) {
1866     
1867         internal_error(true, "RRDSET '%s' switching update every from %d to %d",
1868                        rrdset_id(st), (int)st->update_every, (int)update_every);
1869     
1870         time_t prev_update_every = st->update_every;
>>>     CID 361429:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "update_every" is cast to "int".
1871         st->update_every = update_every;
1872     
1873         // switch update every to the storage engine
1874         RRDDIM *rd;
1875         rrddim_foreach_read(rd, st) {
1876             for (size_t tier = 0; tier < storage_tiers; tier++) {
```

File: streaming/sender.c (hunk 2 above)
Comment (ktsaou) [reply streaming/sender.c:930]: This is intentional, it supports also parsing numbers in hex. I want to experiment with this to improve performance (hex number parsing is much faster than decimal).

File: collectors/plugins.d/pluginsd_parser.c (hunk 3 above)
Comment (ktsaou) [reply collectors/plugins.d/pluginsd_parser.c:931]: This is intentional, it supports also parsing numbers in hex. I want to experiment with this to improve performance (hex number parsing is much faster than decimal).

File: database/rrdset.c (hunk 4 above)
Comment (ktsaou) [reply database/rrdset.c:1870]: `time_t` in 64 bit systems is 64 bits. Generally we need to work on this problem in a more generic way. For moment let's stick to `time_t` and we will fix it later.

**Comment by vkalintiris** [comment]
> I have not tested a proxy yet. If someone can help, please do.

Child with active-active parents

child:
```
[stream]
    enabled = yes
    destination = tcp:localhost:20100 tcp:localhost:20101
    api key = 11111111-2222-3333-4444-555555555555
    timeout seconds = 60
    send charts matching = *
    buffer size bytes = 1048576
    reconnect delay seconds = 5
    initial clock resync iterations = 60
    enable compression = yes

[11111111-2222-3333-4444-555555555555]
    enabled = yes
    enable compression = yes
    allow from = *
```

parent 1:
```
[stream]
    enabled = yes
    destination = tcp:localhost:20101
    api key = 11111111-2222-3333-4444-555555555555
    timeout seconds = 60
    send charts matching = *
    buffer size bytes = 1048576
    reconnect delay seconds = 5
    initial clock resync iterations = 60
    enable compression = yes

[11111111-2222-3333-4444-555555555555]
    enabled = yes
    enable compression = yes
    allow from = *
```

parent 2:
```
[stream]
    enabled = yes
    destination = tcp:localhost:20100
    api key = 11111111-2222-3333-4444-555555555555
    timeout seconds = 60
    send charts matching = *
    buffer size bytes = 1048576
    reconnect delay seconds = 5
    initial clock resync iterations = 60
    enable compression = yes

[11111111-2222-3333-4444-555555555555]
    enabled = yes
    enable compression = yes
    allow from = *
```

1. start child
2. start parent 1
3. start parent 2
4. stop child
5. delete dbengine files of child & start child

Parent 1 has a small gap after the historical data as expected, parent 2 shows the child node active but the dashboard for the child is not being updated. Got:

> Host 'nd0' has history of 0 entries, but the wanted one is 3600 entries. Restart netdata here to apply the new settings.

6. Restart parent 2

Parent 2 does not show any charts for child. Not sure if this test case is meaningful, curious how it'll behave when the storage engine of the child is not dbengine.

**Comment by ktsaou** [comment]
> > Host 'nd0' has history of 0 entries, but the wanted one is 3600 entries. Restart netdata here to apply the new settings.

This is irrelevant (a historical message that is no longer valid at all - this PR didn't touch it).

I will test active-active parents tomorrow.

**Comment by thiagoftsm** [comment]
Hello @ktsaou and @vkalintiris ,

I set up the following scenario:

| Host |  is parent? | memory mode (streaming and host) |
|--------|----------------|-------------------------------------------------|
| Slackware current x86_64 | Yes | ram | 
| FreeBSD x86_64 libvirt | No | dbengine |
|Raspberry OS armvl7 32 | No | dbengine |

With this scenario I did not hope to have the replication working, because parent is not using `dbengine`, but probably the clients and server would fallback to old protocol. Instead I have thousands of messages like these:

```sh
bash-5.2# grep "replication" /var/log/netdata/error.log
2022-10-31 00:33:08: netdata IERR  : STREAM_RECEIVER[freebsd13.localdomain,[192.168.121.242]:27307] : (0233@streaming/replicatio:send_replay_cha): REPLAY: host 'freebsd13.localdomain', chart 'apps.threads': sending replication request 1664235573 [2022-09-26 23:39:33] to 1664236173 [2022-09-26 23:49:33], start streaming: false
2022-10-31 01:16:38: netdata IERR  : STREAM_RECEIVER[potinho,[192.168.1.9]:37490] : (0233@streaming/replicatio:send_replay_cha): REPLAY: host 'potinho', chart 'netdata.dictionaries.other.memory': sending replication request 1667177181 [2022-10-31 00:46:21] to 1667177781 [2022-10-31 00:56:21], start streaming: false
bash-5.2# grep "replication" /var/log/netdata/error.log | wc -l
25672
```

When I have replication disabled on `parent` (`enable replication = no`), I still do not have streaming working and I receive the following messages:

```sh
bash-5.2# grep "replication" /var/log/netdata/error.log
2022-10-31 01:23:59: netdata IERR  : STREAM_RECEIVER[freebsd13.localdomain,[192.168.121.242]:35598] : (0271@streaming/replicatio:replicate_chart): REPLAY: host 'freebsd13.localdomain', chart 'system.cpu': sending empty replication reque
st because replication is disabled
2022-10-31 01:24:52: netdata IERR  : STREAM_RECEIVER[potinho,[192.168.1.9]:46392] : (0271@streaming/replicatio:replicate_chart): REPLAY: host 'potinho', chart 'disk_space._': sending empty replication request because replication is disabled
```

When I start server using current `master` branch and children with the same configuration, I still do not have data:

```sh
2022-10-31 02:25:29: netdata IERR  : WEB_SERVER[static1] : (2814@database/rrdcontext.:query_target_cr): QUERY TARGET: query 'chart://host:potinho/instance:system.cpu/dimensions:|iowait/after:-540/before:0/points:160/group:average/options:nonzero,flip,jsonwrap,ms' does not have all the data required. Matched 1 hosts, 0 contexts, 0 instances, 0 dimensions, 0 metrics to query, 0 metrics skipped because they don't have data in the desired time-frame. Aborting it.
2022-10-31 02:25:30: netdata IERR  : WEB_SERVER[static1] : (2814@database/rrdcontext.:query_target_cr): QUERY TARGET: query 'chart://host:potinho/instance:system.cpu/dimensions:|softirq/after:-540/before:0/points:160/group:average/options:nonzero,flip,jsonwrap,ms' does not have all the data required. Matched 1 hosts, 0 contexts, 0 instances, 0 dimensions, 0 metrics to query, 0 metrics skipped because they don't have data in the desired time-frame. Aborting it.
2022-10-31 02:25:30: netdata IERR  : WEB_SERVER[static1] : (2814@database/rrdcontext.:query_target_cr): QUERY TARGET: query 'chart://host:potinho/instance:system.cpu/dimensions:*/after:-540/before:0/points:392/group:average/options:nonzero,flip,jsonwrap,ms' does not have all the data required. Matched 1 hosts, 0 contexts, 0 instances, 0 dimensions, 0 metrics to query, 0 metrics skipped because they don't have data in the desired time-frame. Aborting it.
2022-10-31 02:25:30: netdata IERR  : WEB_SERVER[static1] : (2814@database/rrdcontext.:query_target_cr): QUERY TARGET: query 'chart://host:potinho/instance:system.swap/dimensions:|used/after:-540/before:0/points:540/group:average/options:nonzero,jsonwrap,abs,percentage' does not have all the data required. Matched 1 hosts, 0 contexts, 0 instances, 0 dimensions, 0 metrics to query, 0 metrics skipped because they don't have data in the desired time-frame. Aborting it.
2022-10-31 02:25:30: netdata IERR  : WEB_SERVER[static1] : (2814@database/rrdcontext.:query_target_cr): QUERY TARGET: query 'chart://host:potinho/instance:system.io/dimensions:|in/after:-540/before:0/points:540/group:average/options:nonzero,jsonwrap,abs' does not have all the data required. Matched 1 hosts, 0 contexts, 0 instances, 0 dimensions, 0 metrics to query, 0 metrics skipped because they don't have data in the desired time-frame. Aborting it.
```

When I tested current `master` on both `child` and `parent` everything works as expected.

Finally when `parent` has current PR with replication disabled and `children` have current `master` the streaming still works.

**Comment by ktsaou** [comment]
I think I solved all the problems:

1. Streaming was broken on proxies when empty replication commands were sent.. Now streaming always resumes.
2. Replication was broken on db modes other than dbengine. Now replication works everywhere.

**Comment by ktsaou** [comment]
> > Host 'nd0' has history of 0 entries, but the wanted one is 3600 entries. Restart netdata here to apply the new settings.

I also pushed a change to avoid printing this on dbengine mode (which is irrelevant anyway).

**Comment by ktsaou** [comment]
And merged master to this again.
This is good to go now...

**Comment by ilyam8** [comment]
@thiagoftsm is this PR related?

> 2022-10-31 13:47:56: netdata ERROR : STREAM_SENDER[potinho] : (0802@streaming/sender.c  :attempt_read   ): STREAM potinho [send to 192.168.1.10:19999] SSL error: error:1408F10B:SSL routines:ssl3_get_record:wrong version number

**Comment by ktsaou** [comment]
> @thiagoftsm is the problem (can't establish a connection) related to this PR?

It shouldn't be. Can you try the same configuration with master? Does it work?

**Comment by ktsaou** [comment]
I think the problem has been introduced with #13720 
The parent just puts data on the wire without passing them through ssl. So, even Functions does not work under SSL streaming connections.

**Comment by thiagoftsm** [comment]
> ssl3_get_record

When I compile master branch on `parent` and `child` I do not have issue on child:

```sh
root@potinho:/home/thiago/Netdata/netdata# grep "ssl3_get_record" /var/log/netdata/error.log 
root@potinho:/home/thiago/Netdata/netdata# grep "stream" /var/log/netdata/error.log 
2022-10-31 15:38:45: netdata INFO  : MAIN : (0130@streaming/rrdpush.c :rrdpush_init   ): Netdata is configured to accept invalid SSL certificate.
2022-10-31 15:38:46: netdata INFO  : MAIN : (0032@streaming/compressio:lz4_compressor_): STREAM_COMPRESSION: Compressor Reset
2022-10-31 15:38:46: netdata INFO  : MAIN : (0533@streaming/rrdpush.c :destinations_in): STREAM: added streaming destination No 1: '192.168.1.10:19999' to host 'potinho'
2022-10-31 15:38:46: netdata INFO  : MAIN : (0451@database/rrdhost.c  :rrdhost_create ): Host 'potinho' (at registry as 'potinho') with guid 'e114e67e-2ac2-11ed-974a-dca632be9f9b' initialized, os 'linux', timezone 'Europe/London', tags '', program_name 'netdata', program_version 'v1.36.0-284-g7b078bc3a', update every 1, memory mode dbengine, history entries 0, streaming enabled (to '192.168.1.10:19999' with api key '293053e2-8a49-4cf5-a32d-616bfaf36128'), health enabled, cache_dir '/var/cache/netdata', varlib_dir '/var/lib/netdata', health_log '(null)', alarms default handler '', alarms default recipient ''
2022-10-31 15:38:47: netdata ERROR : GLOBAL_STATS : (0403@streaming/rrdpush.c :rrdset_done_pus): STREAM potinho [send]: not ready - collected metrics are not sent to parent.
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (1090@streaming/sender.c  :rrdpush_sender_): STREAM potinho [send]: thread created (task id 4529)
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0488@streaming/rrdpush.c :connect_to_one_): STREAM potinho: attempting to connect to '192.168.1.10:19999' (default port: 19999)...
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0407@streaming/sender.c  :rrdpush_sender_): STREAM potinho [send to 192.168.1.10:19999]: initializing communication...
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0611@streaming/sender.c  :rrdpush_sender_): STREAM potinho [send to 192.168.1.10:19999]: waiting response from remote netdata...
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0032@streaming/compressio:lz4_compressor_): STREAM_COMPRESSION: Compressor Reset
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0969@streaming/rrdpush.c :log_sender_capa): STREAM potinho [send to 192.168.1.10:19999]: established link with negotiated capabilities: VCAPS HLABELS CLAIM CLABELS COMPRESSION FUNCTIONS 
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0924@streaming/sender.c  :rrdpush_queue_i): STREAM potinho [send to 192.168.1.10:19999]: sending metric definitions...
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (0932@streaming/sender.c  :rrdpush_queue_i): STREAM potinho [send to 192.168.1.10:19999]: sending metric definitions finished.
2022-10-31 15:38:47: netdata INFO  : STREAM_SENDER[potinho] : (1200@streaming/sender.c  :rrdpush_sender_): STREAM potinho [send to 192.168.1.10:19999]: enabling metrics streaming...
2022-10-31 15:38:47: netdata INFO  : PLUGIN[proc] : (0409@streaming/rrdpush.c :rrdset_done_pus): STREAM potinho [send]: sending metrics to parent...
2022-10-31 15:38:47: go.d ERROR: prometheus[stream_exporter_local] Get "http://127.0.0.1:9178/metrics": dial tcp 127.0.0.1:9178: connect: connection refused
2022-10-31 15:38:47: go.d ERROR: prometheus[stream_exporter_local] check failed
2022-10-31 15:38:47: go.d ERROR: prometheus[openconfig_streaming_telemetry_exporter_local] Get "http://127.0.0.1:9513/metrics": dial tcp 127.0.0.1:9513: connect: connection refused
2022-10-31 15:38:47: go.d ERROR: prometheus[openconfig_streaming_telemetry_exporter_local] check failed
root@potinho:/home/thiago/Netdata/netdata# 
```

And parent receives metric:

```sh
bash-5.2# grep stream /var/log/netdata/error.log 
2022-10-31 15:38:43: netdata INFO  : MAIN : (0451@database/rrdhost.c  :rrdhost_create ): Host 'hades' (at registry as 'hades') with guid 'ab85268a-3ff9-11ed-990a-f82819901f7f' initialized, os 'linux', timezone 'UTC', tags '', program_name 'netdata', program_version 'v1.36.0-290-nightly', update every 1, memory mode ram, history entries 4096, streaming disabled (to '' with api key ''), health enabled, cache_dir '/var/cache/netdata', varlib_dir '/var/lib/netdata', health_log '(null)', alarms default handler '', alarms default recipient ''
2022-10-31 15:38:43: netdata INFO  : MAIN : (0451@database/rrdhost.c  :rrdhost_create ): Host 'potinho' (at registry as 'potinho') with guid 'e114e67e-2ac2-11ed-974a-dca632be9f9b' initialized, os 'linux', timezone 'Europe/London', tags '', program_name 'netdata', program_version 'v1.36.0-284-g7b078bc3a', update every 1, memory mode ram, history entries 4096, streaming disabled (to '' with api key ''), health disabled, cache_dir '/var/cache/netdata/e114e67e-2ac2-11ed-974a-dca632be9f9b', varlib_dir '/var/lib/netdata/e114e67e-2ac2-11ed-974a-dca632be9f9b', health_log '(null)', alarms default handler '', alarms default recipient ''
2022-10-31 15:38:47: netdata INFO  : WEB_SERVER[static3] : (0647@streaming/rrdpush.c :rrdpush_receive): clients wants to STREAM metrics.
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0816@streaming/receiver.c:rrdpush_receive): STREAM potinho [192.168.1.9]:50428: receive thread created (task id 24944)
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0451@database/rrdhost.c  :rrdhost_create ): Host 'potinho' (at registry as 'potinho') with guid 'e114e67e-2ac2-11ed-974a-dca632be9f9b' initialized, os 'linux', timezone 'Europe/London', tags '', program_name 'netdata', program_version 'v1.36.0-284-g7b078bc3a', update every 1, memory mode ram, history entries 4096, streaming disabled (to '' with api key ''), health enabled, cache_dir '/var/cache/netdata/e114e67e-2ac2-11ed-974a-dca632be9f9b', varlib_dir '/var/lib/netdata/e114e67e-2ac2-11ed-974a-dca632be9f9b', health_log '(null)', alarms default handler '', alarms default recipient ''
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0614@streaming/receiver.c:rrdpush_receive): STREAM potinho [receive from [192.168.1.9]:50428]: client willing to stream metrics for host 'potinho' with machine_guid 'e114e67e-2ac2-11ed-974a-dca632be9f9b': update every = 1, history = 4096, memory mode = ram, health auto, SSL, tags ''
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0655@streaming/receiver.c:rrdpush_receive): STREAM potinho [receive from [192.168.1.9]:50428]: initializing communication...
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0959@streaming/rrdpush.c :log_receiver_ca): STREAM potinho [receive from [192.168.1.9]:50428]: established link with negotiated capabilities: VCAPS HLABELS CLAIM CLABELS COMPRESSION FUNCTIONS 
2022-10-31 15:38:47: netdata INFO  : STREAM_RECEIVER[potinho,[192.168.1.9]:50428] : (0751@streaming/receiver.c:rrdpush_receive): STREAM potinho [receive from [192.168.1.9]:50428]: receiving metrics...
```

Screenshot_20221031_154330


So we cannot discard the relationship.

**Comment by thiagoftsm** [comment]
> > @thiagoftsm is the problem (can't establish a connection) related to this PR?
> 
> It shouldn't be. Can you try the same configuration with master? Does it work?

Yes, everything is working when I have master on `parent` and `child` as I showed https://github.com/netdata/netdata/pull/13873#issuecomment-1297285963

**Comment by thiagoftsm** [comment]
@ktsaou after two last commits I recompiled `parent` with this branch and I kept child with current master`, and now for the first time I have streaming working in the cited scenario.

I will proceed with tests and WMI development. :handshake:

**Review by ktsaou — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
After last commits all tests passed with different memory modes.
I did not test `dbengine`, because this is the main `memory mode`, so I focus the others.

**Review by ilyam8 — APPROVED** [review]

**Review by stelfrag — APPROVED** [review]

**Comment by amoss** [comment]
Nice work 🎉 Good to see this working

**Pull request merged**